reference/test/matrix/ell_kernels.cpp

```diff
@@ -0,0 +1,346 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright 2017-2018
+
+Karlsruhe Institute of Technology
+Universitat Jaume I
+University of Tennessee
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <core/matrix/ell.hpp>
+
+
+#include <gtest/gtest.h>
+
+
+#include <core/base/exception.hpp>
+#include <core/base/exception_helpers.hpp>
+#include <core/base/executor.hpp>
+#include <core/matrix/dense.hpp>
+#include <core/test/utils/assertions.hpp>
+
+
+namespace {
+
+
+class Ell : public ::testing::Test {
+protected:
+    using Mtx = gko::matrix::Ell<>;
+    using Vec = gko::matrix::Dense<>;
+
+    Ell()
+        : exec(gko::ReferenceExecutor::create()),
+          mtx1(Mtx::create(exec, 2, 3, 3, 2)),
+          mtx2(Mtx::create(exec, 2, 3, 3, 16))
+    {
+        Mtx::value_type *v = mtx1->get_values();
+        Mtx::index_type *c = mtx1->get_col_idxs();
+        Mtx::index_type n = mtx1->get_max_nonzeros_per_row();
+        c[0] = 0;
+        c[1] = 1;
+        c[2] = 1;
+        c[3] = 0;
+        c[4] = 2;
+        c[5] = 0;
+        v[0] = 1.0;
+        v[1] = 5.0;
+        v[2] = 3.0;
+        v[3] = 0.0;
+        v[4] = 2.0;
+        v[5] = 0.0;
```

Comment:
I'm thinking what's the best way to do this thing with test matrix generation. In general, we discussed this problem recently in the private repo (I hope to soon have contributing guidelines where we summarize all such decisions, and update the existing unit tests accordingly), and we concluded that if we are explicitly setting the matrix values like here, there should be a comment above which formats the matrix nicely:

```c++
/*
  1 0 0
  0 5 2
  0 3 0 
*/
c[0] = 0;
...
```

In general, this is error prone (e.g. there could be a discrepancy between what the comment says an what is actually being set), so the preferred way of doing it is by using `gko::initialize`:

```c++
// clang-format off
auto mtx1 = gko::initialize<Mtx>(
    {{1.0, 0.0, 0.0},
     {0.0, 5.0, 2.0},
     {0.0, 3.0, 0.0}), exec);
// clang-format on
```

This approach eliminates the possibility of errors the other approach has, but it does rely on a working conversion between Dense and the format you are testing. For this reason always use the `gko::initialize` approach, __except__ when testing conversion from/to Dense.

So in this particular case, I would initialize the matrices here with `gko::initialize` (and implement the `ConvertibleTo<Ell>` interface in `Dense`), but use explicitly initialized matrices when testing conversions from and to dense.

Reply:
@gflegar 
I am not sure whether I understand what you mean.
I implement two methods for Dense to ELL. (padding or no padding)
I can use initialize for Dense to ELL with no padding, but I use move_to/convert_to not initialize for Dense to ELL with user-defined padding.
Is there anything I misunderstand?
Thanks.

Reply:
I think I understand what the problem is: when you use `Dense::move_to(Ell *result)` or `Dense::convert_to(Ell *result)`, how do you know what padding to set for `result`?
This is an interesting question.

The current solution you have is to pass an extra parameter to set this padding. However, the problem with this solution is that the method is not a part of `ConvertibleTo<Ell>`, so it won't be visible when using `Dense` through `ConvertibleTo<Ell>`.
On the other hand, we do want the user to be able to specify the padding when converting to Ell. To be able to do both - have the library select the padding automatically, or have the user select it by themselves, I would suggest the following:

We only have one implementation of `convert_to(Ell *result)`/ `move_to(Ell *result)`, which takes only the matrix as input  (i.e. the one that implements `ConvertibleTo<Ell>` interface), and we implement it as follows:

*   If the padding of `result` is a valid padding for the current matrix, we keep this padding
*   otherwise, we change the padding to the smallest valid value (i.e. number of rows of the matrix)

This way, we can use `initialize` in both cases:

```c++
// uses the Ell(exec), to construct the result,
// so move_to will set the padding to smallest possible value (i.e. 2)
auto mtx = initialize<Ell>({{1, 2}, {0, 1}}, exec);
// uses the Ell(exec, 0, 0, 3) to construct the matrix,
// so move_to will keep the padding set to 3
auto mtx2 = initialize<Ell>({{1, 2}, {0, 1}}, exec, 0, 0, 3);
```

What do you think about this solution?

Reply:
@gflegar 
This solution is good for me.
I have done it
Users can also define max_nonzeros_per_row and padding now.